src/compiler.ts

```diff
@@ -815,6 +815,13 @@ export class Compiler extends DiagnosticEmitter {
         if (!classInstance.type.isUnmanaged) {
           let module = this.module;
           let internalName = classInstance.internalName;
+
+          // We remove the global before adding it, as this ensures
+          // we are not adding the same global twice. This is needed,
+          // As we can export a class from a file, and in that sample file,
+          // Export a namspace (import * as), that exports the same class.
+          // And there is no module.hasGlobal(), and this works fine :)
+          module.removeGlobal(internalName);
```

Comment:
Iiuc the problem here is that the same class can be exported using two different names, with the simplest case being

```ts
export class Foo {}
export { Foo as Bar }
```

so we get the exports

```
Foo = index/Foo
Bar = index/Foo
```

That correct?

Regarding removing and re-adding the global, this is a little dangerous because Binaryen APIs typically error when operating on something that doesn't exist. I can confirm that it doesn't error on `BinaryenRemoveGlobal`, but this might be an oversight with a slight chance to change in the future. As an alternative, I'd suggest to introduce a new `Compiler#exportedClasses: Set<Class>` and only add the global if it isn't present in the set yet.

Reply:
> Iiuc the problem here is that the same class can be exported using two different names

Ah not quite. The problem is if we have the code:

**one.ts**
```
export class MyClass {}
```

**two.ts**
```
export {MyClass} from "./one.ts"
import * as MyNamespace from "./one.ts"
export {MyNamespace}
```

When we traverse this part of the code. The first export of MyClass will add it as a global. But then we we export the namespace, this will go and traverse the name space members, which also contains MyClass, and thus will try to add it as a global. And then Binaryen will error out because it's already been created as a global, thus creating a conflict (when really it's the exact same thing).

> Regarding removing and re-adding the global, this is a little dangerous because Binaryen APIs typically error when operating on something that doesn't exist. I can confirm that it doesn't error on BinaryenRemoveGlobal, but this might be an oversight with a slight chance to change in the future. As an alternative, I'd suggest to introduce a new Compiler#exportedClasses: Set<Class> and only add the global if it isn't present in the set yet.

Oh yeah I'm totally down to not do this in my hack-y way and that totally makes sense haha! :joy: 

And just to be clear, I assume `Compiler#exportedClasses: Set<Class>` is referring to `src/compiler.ts`, https://github.com/AssemblyScript/assemblyscript/blob/master/src/compiler.ts#L327 and not something in Binaryen correct? :smile:

Reply:
Right, just a set on the `Compiler` class to keep track of the classes already handled, so the implicit id global is only added if it doesn't yet exist :). Also wondering if anyone ever used the implicit id, as it's really just a convenience shortcut so one doesn't have to

```ts
export const MyClassID = idof<MyClass>;
```

when already

```ts
export { MyClass };
```

so externally one can do `exports.MyClass` to obtain the id instead of the manual `exports.MyClassID`.

Reply:
> Also wondering if anyone ever used the implicit id, as it's really just a convenience shortcut so one doesn't have to

I'm not too sure as I've really just been focused on the importing/exporting between files here :joy: But it does seem like a nice-to-have/nice-to-try later on :smile: